File: 01-libsql-vs-duckdb/benchmark.py
#!/usr/bin/env python3
"""
libSQL vs DuckDB ベンチマーク
大量データでのパフォーマンス比較
"""

import sqlite3
import duckdb
import time
import random
import string
from pathlib import Path


def generate_random_data(n):
    """ランダムなテストデータを生成"""
    data = []
    for i in range(n):
        name = ''.join(random.choices(string.ascii_letters, k=10))
        email = f"{name.lower()}@example.com"
        age = random.randint(18, 80)
        salary = random.randint(30000, 150000)
        department = random.choice(['Engineering', 'Sales', 'Marketing', 'HR', 'Finance'])
        data.append((i + 1, name, email, age, salary, department))
    return data


def benchmark_libsql(data_size):
    """libSQL (SQLite) のベンチマーク"""
    print(f"\n{'='*60}")
    print(f"libSQL ベンチマーク (データサイズ: {data_size:,}件)")
    print('='*60)

    db_path = "benchmark_libsql.db"
    Path(db_path).unlink(missing_ok=True)

    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()

    # テーブル作成
    cursor.execute("""
        CREATE TABLE employees (
            id INTEGER PRIMARY KEY,
            name TEXT NOT NULL,
            email TEXT NOT NULL,
            age INTEGER,
            salary INTEGER,
            department TEXT
        )
    """)

    # インデックス作成
    cursor.execute("CREATE INDEX idx_department ON employees(department)")
    cursor.execute("CREATE INDEX idx_age ON employees(age)")

    # データ挿入（バッチ処理）
    print("\n[1] データ挿入テスト")
    data = generate_random_data(data_size)
    start = time.time()
    cursor.executemany(
        "INSERT INTO employees VALUES (?, ?, ?, ?, ?, ?)",
        data
    )
    conn.commit()
    insert_time = time.time() - start
    print(f"  ✓ {data_size:,}件の挿入: {insert_time:.3f}秒 ({data_size/insert_time:.0f}件/秒)")

    # 単純なSELECT
    print("\n[2] 単純なSELECTクエリ")
    start = time.time()
    cursor.execute("SELECT * FROM employees WHERE age > 50 LIMIT 100")
    results = cursor.fetchall()
    select_time = time.time() - start
    print(f"  ✓ 条件付きSELECT: {select_time*1000:.2f}ms ({len(results)}件)")

    # 集計クエリ
    print("\n[3] 集計クエリ")
    start = time.time()
    cursor.execute("""
        SELECT department, COUNT(*) as count, AVG(salary) as avg_salary
        FROM employees
        GROUP BY department
        ORDER BY avg_salary DESC
    """)
    results = cursor.fetchall()
    agg_time = time.time() - start
    print(f"  ✓ GROUP BY集計: {agg_time*1000:.2f}ms")
    for row in results[:3]:
        print(f"    - {row[0]}: {row[1]}人, 平均年収 ${row[2]:,.0f}")

    # JOIN相当のクエリ（自己結合）
    print("\n[4] 複雑な集計クエリ")
    start = time.time()
    cursor.execute("""
        SELECT
            department,
            AVG(age) as avg_age,
            AVG(salary) as avg_salary,
            COUNT(*) as count
        FROM employees
        WHERE salary > 50000
        GROUP BY department
        HAVING COUNT(*) > 10
        ORDER BY avg_salary DESC
    """)
    results = cursor.fetchall()
    complex_time = time.time() - start
    print(f"  ✓ 複雑な集計: {complex_time*1000:.2f}ms ({len(results)}部署)")

    # UPDATE
    print("\n[5] UPDATEクエリ (トランザクション)")
    start = time.time()
    cursor.execute("UPDATE employees SET salary = salary * 1.05 WHERE department = 'Engineering'")
    conn.commit()
    update_time = time.time() - start
    updated = cursor.rowcount
    print(f"  ✓ UPDATE: {update_time*1000:.2f}ms ({updated}件更新)")

    conn.close()

    return {
        "insert": insert_time,
        "select": select_time,
        "aggregate": agg_time,
        "complex": complex_time,
        "update": update_time
    }


def benchmark_duckdb(data_size):
    """DuckDB のベンチマーク"""
    print(f"\n{'='*60}")
    print(f"DuckDB ベンチマーク (データサイズ: {data_size:,}件)")
    print('='*60)

    db_path = "benchmark_duckdb.duckdb"
    Path(db_path).unlink(missing_ok=True)

    conn = duckdb.connect(db_path)

    # テーブル作成
    conn.execute("""
        CREATE TABLE employees (
            id INTEGER PRIMARY KEY,
            name VARCHAR NOT NULL,
            email VARCHAR NOT NULL,
            age INTEGER,
            salary INTEGER,
            department VARCHAR
        )
    """)

    # インデックス作成
    conn.execute("CREATE INDEX idx_department ON employees(department)")
    conn.execute("CREATE INDEX idx_age ON employees(age)")

    # データ挿入（バッチ処理）
    print("\n[1] データ挿入テスト")
    data = generate_random_data(data_size)
    start = time.time()
    conn.executemany(
        "INSERT INTO employees VALUES (?, ?, ?, ?, ?, ?)",
        data
    )
    insert_time = time.time() - start
    print(f"  ✓ {data_size:,}件の挿入: {insert_time:.3f}秒 ({data_size/insert_time:.0f}件/秒)")

    # 単純なSELECT
    print("\n[2] 単純なSELECTクエリ")
    start = time.time()
    results = conn.execute("SELECT * FROM employees WHERE age > 50 LIMIT 100").fetchall()
    select_time = time.time() - start
    print(f"  ✓ 条件付きSELECT: {select_time*1000:.2f}ms ({len(results)}件)")

    # 集計クエリ（DuckDBの強み）
    print("\n[3] 集計クエリ (DuckDBの強み)")
    start = time.time()
    results = conn.execute("""
        SELECT department, COUNT(*) as count, AVG(salary) as avg_salary
        FROM employees
        GROUP BY department
        ORDER BY avg_salary DESC
    """).fetchall()
    agg_time = time.time() - start
    print(f"  ✓ GROUP BY集計: {agg_time*1000:.2f}ms")
    for row in results[:3]:
        print(f"    - {row[0]}: {row[1]}人, 平均年収 ${row[2]:,.0f}")

    # 複雑な集計クエリ
    print("\n[4] 複雑な集計クエリ (DuckDBの強み)")
    start = time.time()
    results = conn.execute("""
        SELECT
            department,
            AVG(age) as avg_age,
            AVG(salary) as avg_salary,
            COUNT(*) as count
        FROM employees
        WHERE salary > 50000
        GROUP BY department
        HAVING COUNT(*) > 10
        ORDER BY avg_salary DESC
    """).fetchall()
    complex_time = time.time() - start
    print(f"  ✓ 複雑な集計: {complex_time*1000:.2f}ms ({len(results)}部署)")

    # ウィンドウ関数（DuckDBの高度な機能）
    print("\n[5] ウィンドウ関数 (DuckDBの高度な機能)")
    start = time.time()
    results = conn.execute("""
        SELECT
            department,
            name,
            salary,
            AVG(salary) OVER (PARTITION BY department) as dept_avg_salary,
            RANK() OVER (PARTITION BY department ORDER BY salary DESC) as rank_in_dept
        FROM employees
        WHERE department IN ('Engineering', 'Sales')
        LIMIT 10
    """).fetchall()
    window_time = time.time() - start
    print(f"  ✓ ウィンドウ関数: {window_time*1000:.2f}ms")

    # UPDATE
    print("\n[6] UPDATEクエリ")
    start = time.time()
    conn.execute("UPDATE employees SET salary = salary * 1.05 WHERE department = 'Engineering'")
    update_time = time.time() - start
    print(f"  ✓ UPDATE: {update_time*1000:.2f}ms")

    conn.close()

    return {
        "insert": insert_time,
        "select": select_time,
        "aggregate": agg_time,
        "complex": complex_time,
        "window": window_time,
        "update": update_time
    }


def print_comparison(libsql_times, duckdb_times, data_size):
    """ベンチマーク結果の比較表示"""
    print(f"\n{'='*60}")
    print(f"ベンチマーク結果比較 (データサイズ: {data_size:,}件)")
    print('='*60)

    print(f"\n{'操作':<20} {'libSQL':>15} {'DuckDB':>15} {'高速':<10} {'差分'}")
    print('-' * 70)

    # 挿入
    libsql_insert = libsql_times['insert']
    duckdb_insert = duckdb_times['insert']
    faster = "libSQL" if libsql_insert < duckdb_insert else "DuckDB"
    ratio = max(libsql_insert, duckdb_insert) / min(libsql_insert, duckdb_insert)
    print(f"{'データ挿入':<20} {libsql_insert:>12.3f}秒 {duckdb_insert:>12.3f}秒 {faster:<10} {ratio:.2f}x")

    # SELECT
    libsql_select = libsql_times['select'] * 1000
    duckdb_select = duckdb_times['select'] * 1000
    faster = "libSQL" if libsql_select < duckdb_select else "DuckDB"
    ratio = max(libsql_select, duckdb_select) / min(libsql_select, duckdb_select)
    print(f"{'SELECT':<20} {libsql_select:>13.2f}ms {duckdb_select:>13.2f}ms {faster:<10} {ratio:.2f}x")

    # 集計
    libsql_agg = libsql_times['aggregate'] * 1000
    duckdb_agg = duckdb_times['aggregate'] * 1000
    faster = "libSQL" if libsql_agg < duckdb_agg else "DuckDB"
    ratio = max(libsql_agg, duckdb_agg) / min(libsql_agg, duckdb_agg)
    print(f"{'GROUP BY集計':<20} {libsql_agg:>13.2f}ms {duckdb_agg:>13.2f}ms {faster:<10} {ratio:.2f}x")

    # 複雑な集計
    libsql_complex = libsql_times['complex'] * 1000
    duckdb_complex = duckdb_times['complex'] * 1000
    faster = "libSQL" if libsql_complex < duckdb_complex else "DuckDB"
    ratio = max(libsql_complex, duckdb_complex) / min(libsql_complex, duckdb_complex)
    print(f"{'複雑な集計':<20} {libsql_complex:>13.2f}ms {duckdb_complex:>13.2f}ms {faster:<10} {ratio:.2f}x")

    # UPDATE
    libsql_update = libsql_times['update'] * 1000
    duckdb_update = duckdb_times['update'] * 1000
    faster = "libSQL" if libsql_update < duckdb_update else "DuckDB"
    ratio = max(libsql_update, duckdb_update) / min(libsql_update, duckdb_update)
    print(f"{'UPDATE':<20} {libsql_update:>13.2f}ms {duckdb_update:>13.2f}ms {faster:<10} {ratio:.2f}x")

    print(f"\n{'='*60}")
    print("結論")
    print('='*60)
    print("\n【libSQL】")
    print("  ✓ トランザクション処理 (OLTP) に最適")
    print("  ✓ 小〜中規模データでの高速な読み書き")
    print("  ✓ 行指向ストレージで個別レコード操作が得意")
    print("  ✓ 組み込みアプリケーションに最適")

    print("\n【DuckDB】")
    print("  ✓ 分析クエリ (OLAP) に最適")
    print("  ✓ 大量データの集計・グループ化が高速")
    print("  ✓ カラムナストレージによる効率的なスキャン")
    print("  ✓ ウィンドウ関数などの高度な分析機能")
    print("  ✓ データサイエンス・BI用途に最適")


def main():
    print("="*60)
    print("libSQL vs DuckDB パフォーマンスベンチマーク")
    print("="*60)

    # データサイズを選択
    print("\nベンチマークのデータサイズを選択してください:")
    print("  1. 小 (10,000件) - 高速")
    print("  2. 中 (50,000件) - 推奨")
    print("  3. 大 (100,000件) - 時間がかかります")

    try:
        choice = input("\n選択 (1-3, デフォルト=2): ").strip()
        if not choice:
            choice = "2"

        data_sizes = {"1": 10000, "2": 50000, "3": 100000}
        data_size = data_sizes.get(choice, 50000)

        print(f"\n{data_size:,}件のデータでベンチマークを開始します...")

        # ベンチマーク実行
        libsql_times = benchmark_libsql(data_size)
        duckdb_times = benchmark_duckdb(data_size)

        # 結果比較
        print_comparison(libsql_times, duckdb_times, data_size)

        # クリーンアップ
        print("\n\nクリーンアップ中...")
        Path("benchmark_libsql.db").unlink(missing_ok=True)
        Path("benchmark_duckdb.duckdb").unlink(missing_ok=True)
        Path("benchmark_duckdb.duckdb.wal").unlink(missing_ok=True)
        print("完了!")

    except KeyboardInterrupt:
        print("\n\nベンチマークが中断されました。")
        return 1
    except Exception as e:
        print(f"\nエラーが発生しました: {e}")
        import traceback
        traceback.print_exc()
        return 1

    return 0


if __name__ == "__main__":
    import sys
    sys.exit(main())

File: 01-libsql-vs-duckdb/src/benchmark.ts
#!/usr/bin/env node
/**
 * libSQL vs DuckDB ベンチマーク
 * 大量データでのパフォーマンス比較
 */

import { createClient } from '@libsql/client';
import duckdb from 'duckdb';
import { unlink } from 'fs/promises';
import { existsSync } from 'fs';
import * as readline from 'readline';

interface Employee {
  id: number;
  name: string;
  email: string;
  age: number;
  salary: number;
  department: string;
}

interface BenchmarkResults {
  insert: number;
  select: number;
  aggregate: number;
  complex: number;
  update: number;
  window?: number;
}

function generateRandomData(n: number): Employee[] {
  const departments = ['Engineering', 'Sales', 'Marketing', 'HR', 'Finance'];
  const data: Employee[] = [];

  for (let i = 0; i < n; i++) {
    const name = Math.random().toString(36).substring(2, 12);
    const email = `${name}@example.com`;
    const age = Math.floor(Math.random() * (80 - 18 + 1)) + 18;
    const salary = Math.floor(Math.random() * (150000 - 30000 + 1)) + 30000;
    const department = departments[Math.floor(Math.random() * departments.length)];

    data.push({
      id: i + 1,
      name,
      email,
      age,
      salary,
      department,
    });
  }

  return data;
}

async function benchmarkLibSQL(dataSize: number): Promise<BenchmarkResults> {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`libSQL ベンチマーク (データサイズ: ${dataSize.toLocaleString()}件)`);
  console.log('='.repeat(60));

  const dbPath = 'benchmark_libsql.db';
  if (existsSync(dbPath)) {
    await unlink(dbPath);
  }

  const client = createClient({ url: `file:${dbPath}` });

  // テーブル作成
  await client.execute(`
    CREATE TABLE employees (
      id INTEGER PRIMARY KEY,
      name TEXT NOT NULL,
      email TEXT NOT NULL,
      age INTEGER,
      salary INTEGER,
      department TEXT
    )
  `);

  // インデックス作成
  await client.execute('CREATE INDEX idx_department ON employees(department)');
  await client.execute('CREATE INDEX idx_age ON employees(age)');

  // データ挿入（バッチ処理）
  console.log('\n[1] データ挿入テスト');
  const data = generateRandomData(dataSize);
  const insertStart = performance.now();

  for (const emp of data) {
    await client.execute({
      sql: 'INSERT INTO employees VALUES (?, ?, ?, ?, ?, ?)',
      args: [emp.id, emp.name, emp.email, emp.age, emp.salary, emp.department],
    });
  }

  const insertTime = (performance.now() - insertStart) / 1000;
  console.log(`  ✓ ${dataSize.toLocaleString()}件の挿入: ${insertTime.toFixed(3)}秒 (${Math.floor(dataSize / insertTime).toLocaleString()}件/秒)`);

  // 単純なSELECT
  console.log('\n[2] 単純なSELECTクエリ');
  const selectStart = performance.now();
  const result = await client.execute('SELECT * FROM employees WHERE age > 50 LIMIT 100');
  const selectTime = (performance.now() - selectStart) / 1000;
  console.log(`  ✓ 条件付きSELECT: ${(selectTime * 1000).toFixed(2)}ms (${result.rows.length}件)`);

  // 集計クエリ
  console.log('\n[3] 集計クエリ');
  const aggStart = performance.now();
  const aggResult = await client.execute(`
    SELECT department, COUNT(*) as count, AVG(salary) as avg_salary
    FROM employees
    GROUP BY department
    ORDER BY avg_salary DESC
  `);
  const aggTime = (performance.now() - aggStart) / 1000;
  console.log(`  ✓ GROUP BY集計: ${(aggTime * 1000).toFixed(2)}ms`);
  for (let i = 0; i < Math.min(3, aggResult.rows.length); i++) {
    const row = aggResult.rows[i];
    console.log(`    - ${row.department}: ${row.count}人, 平均年収 $${Number(row.avg_salary).toLocaleString(undefined, {maximumFractionDigits: 0})}`);
  }

  // 複雑な集計クエリ
  console.log('\n[4] 複雑な集計クエリ');
  const complexStart = performance.now();
  const complexResult = await client.execute(`
    SELECT
      department,
      AVG(age) as avg_age,
      AVG(salary) as avg_salary,
      COUNT(*) as count
    FROM employees
    WHERE salary > 50000
    GROUP BY department
    HAVING COUNT(*) > 10
    ORDER BY avg_salary DESC
  `);
  const complexTime = (performance.now() - complexStart) / 1000;
  console.log(`  ✓ 複雑な集計: ${(complexTime * 1000).toFixed(2)}ms (${complexResult.rows.length}部署)`);

  // UPDATE
  console.log('\n[5] UPDATEクエリ (トランザクション)');
  const updateStart = performance.now();
  await client.execute("UPDATE employees SET salary = salary * 1.05 WHERE department = 'Engineering'");
  const updateTime = (performance.now() - updateStart) / 1000;
  console.log(`  ✓ UPDATE: ${(updateTime * 1000).toFixed(2)}ms`);

  // クリーンアップ
  if (existsSync(dbPath)) {
    await unlink(dbPath);
  }

  return {
    insert: insertTime,
    select: selectTime,
    aggregate: aggTime,
    complex: complexTime,
    update: updateTime,
  };
}

async function benchmarkDuckDB(dataSize: number): Promise<BenchmarkResults> {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`DuckDB ベンチマーク (データサイズ: ${dataSize.toLocaleString()}件)`);
  console.log('='.repeat(60));

  const dbPath = 'benchmark_duckdb.duckdb';
  if (existsSync(dbPath)) {
    await unlink(dbPath);
  }

  return new Promise((resolve, reject) => {
    const db = new duckdb.Database(dbPath);
    const conn = db.connect();

    const results: BenchmarkResults = {
      insert: 0,
      select: 0,
      aggregate: 0,
      complex: 0,
      update: 0,
      window: 0,
    };

    conn.run(`
      CREATE TABLE employees (
        id INTEGER PRIMARY KEY,
        name VARCHAR NOT NULL,
        email VARCHAR NOT NULL,
        age INTEGER,
        salary INTEGER,
        department VARCHAR
      )
    `, (err) => {
      if (err) {
        reject(err);
        return;
      }

      // インデックス作成
      conn.run('CREATE INDEX idx_department ON employees(department)', (err) => {
        if (err) {
          reject(err);
          return;
        }

        conn.run('CREATE INDEX idx_age ON employees(age)', async (err) => {
          if (err) {
            reject(err);
            return;
          }

          // データ挿入
          console.log('\n[1] データ挿入テスト');
          const data = generateRandomData(dataSize);
          const insertStart = performance.now();

          let insertCount = 0;
          for (const emp of data) {
            conn.run(
              'INSERT INTO employees VALUES (?, ?, ?, ?, ?, ?)',
              emp.id,
              emp.name,
              emp.email,
              emp.age,
              emp.salary,
              emp.department,
              (err) => {
                if (err) {
                  reject(err);
                  return;
                }
                insertCount++;

                if (insertCount === data.length) {
                  results.insert = (performance.now() - insertStart) / 1000;
                  console.log(`  ✓ ${dataSize.toLocaleString()}件の挿入: ${results.insert.toFixed(3)}秒 (${Math.floor(dataSize / results.insert).toLocaleString()}件/秒)`);

                  // SELECT
                  console.log('\n[2] 単純なSELECTクエリ');
                  const selectStart = performance.now();
                  conn.all('SELECT * FROM employees WHERE age > 50 LIMIT 100', (err, rows) => {
                    if (err) {
                      reject(err);
                      return;
                    }
                    results.select = (performance.now() - selectStart) / 1000;
                    console.log(`  ✓ 条件付きSELECT: ${(results.select * 1000).toFixed(2)}ms (${rows.length}件)`);

                    // 集計
                    console.log('\n[3] 集計クエリ (DuckDBの強み)');
                    const aggStart = performance.now();
                    conn.all(`
                      SELECT department, COUNT(*) as count, AVG(salary) as avg_salary
                      FROM employees
                      GROUP BY department
                      ORDER BY avg_salary DESC
                    `, (err, rows) => {
                      if (err) {
                        reject(err);
                        return;
                      }
                      results.aggregate = (performance.now() - aggStart) / 1000;
                      console.log(`  ✓ GROUP BY集計: ${(results.aggregate * 1000).toFixed(2)}ms`);
                      for (let i = 0; i < Math.min(3, rows.length); i++) {
                        const row = rows[i];
                        console.log(`    - ${row.department}: ${row.count}人, 平均年収 $${Number(row.avg_salary).toLocaleString(undefined, {maximumFractionDigits: 0})}`);
                      }

                      // 複雑な集計
                      console.log('\n[4] 複雑な集計クエリ (DuckDBの強み)');
                      const complexStart = performance.now();
                      conn.all(`
                        SELECT
                          department,
                          AVG(age) as avg_age,
                          AVG(salary) as avg_salary,
                          COUNT(*) as count
                        FROM employees
                        WHERE salary > 50000
                        GROUP BY department
                        HAVING COUNT(*) > 10
                        ORDER BY avg_salary DESC
                      `, (err, rows) => {
                        if (err) {
                          reject(err);
                          return;
                        }
                        results.complex = (performance.now() - complexStart) / 1000;
                        console.log(`  ✓ 複雑な集計: ${(results.complex * 1000).toFixed(2)}ms (${rows.length}部署)`);

                        // ウィンドウ関数
                        console.log('\n[5] ウィンドウ関数 (DuckDBの高度な機能)');
                        const windowStart = performance.now();
                        conn.all(`
                          SELECT
                            department,
                            name,
                            salary,
                            AVG(salary) OVER (PARTITION BY department) as dept_avg_salary,
                            RANK() OVER (PARTITION BY department ORDER BY salary DESC) as rank_in_dept
                          FROM employees
                          WHERE department IN ('Engineering', 'Sales')
                          LIMIT 10
                        `, (err, rows) => {
                          if (err) {
                            reject(err);
                            return;
                          }
                          results.window = (performance.now() - windowStart) / 1000;
                          console.log(`  ✓ ウィンドウ関数: ${(results.window * 1000).toFixed(2)}ms`);

                          // UPDATE
                          console.log('\n[6] UPDATEクエリ');
                          const updateStart = performance.now();
                          conn.run("UPDATE employees SET salary = salary * 1.05 WHERE department = 'Engineering'", async (err) => {
                            if (err) {
                              reject(err);
                              return;
                            }
                            results.update = (performance.now() - updateStart) / 1000;
                            console.log(`  ✓ UPDATE: ${(results.update * 1000).toFixed(2)}ms`);

                            db.close();

                            // クリーンアップ
                            if (existsSync(dbPath)) {
                              await unlink(dbPath);
                            }
                            const walPath = `${dbPath}.wal`;
                            if (existsSync(walPath)) {
                              await unlink(walPath);
                            }

                            resolve(results);
                          });
                        });
                      });
                    });
                  });
                }
              }
            );
          }
        });
      });
    });
  });
}

function printComparison(libsqlTimes: BenchmarkResults, duckdbTimes: BenchmarkResults, dataSize: number): void {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`ベンチマーク結果比較 (データサイズ: ${dataSize.toLocaleString()}件)`);
  console.log('='.repeat(60));

  console.log(`\n${'操作'.padEnd(20)} ${'libSQL'.padStart(15)} ${'DuckDB'.padStart(15)} ${'高速'.padEnd(10)} 差分`);
  console.log('-'.repeat(70));

  // 挿入
  const faster1 = libsqlTimes.insert < duckdbTimes.insert ? 'libSQL' : 'DuckDB';
  const ratio1 = Math.max(libsqlTimes.insert, duckdbTimes.insert) / Math.min(libsqlTimes.insert, duckdbTimes.insert);
  console.log(`${'データ挿入'.padEnd(20)} ${(libsqlTimes.insert + '秒').padStart(15)} ${(duckdbTimes.insert + '秒').padStart(15)} ${faster1.padEnd(10)} ${ratio1.toFixed(2)}x`);

  // SELECT
  const libsqlSelect = libsqlTimes.select * 1000;
  const duckdbSelect = duckdbTimes.select * 1000;
  const faster2 = libsqlSelect < duckdbSelect ? 'libSQL' : 'DuckDB';
  const ratio2 = Math.max(libsqlSelect, duckdbSelect) / Math.min(libsqlSelect, duckdbSelect);
  console.log(`${'SELECT'.padEnd(20)} ${(libsqlSelect.toFixed(2) + 'ms').padStart(15)} ${(duckdbSelect.toFixed(2) + 'ms').padStart(15)} ${faster2.padEnd(10)} ${ratio2.toFixed(2)}x`);

  // 集計
  const libsqlAgg = libsqlTimes.aggregate * 1000;
  const duckdbAgg = duckdbTimes.aggregate * 1000;
  const faster3 = libsqlAgg < duckdbAgg ? 'libSQL' : 'DuckDB';
  const ratio3 = Math.max(libsqlAgg, duckdbAgg) / Math.min(libsqlAgg, duckdbAgg);
  console.log(`${'GROUP BY集計'.padEnd(20)} ${(libsqlAgg.toFixed(2) + 'ms').padStart(15)} ${(duckdbAgg.toFixed(2) + 'ms').padStart(15)} ${faster3.padEnd(10)} ${ratio3.toFixed(2)}x`);

  // 複雑な集計
  const libsqlComplex = libsqlTimes.complex * 1000;
  const duckdbComplex = duckdbTimes.complex * 1000;
  const faster4 = libsqlComplex < duckdbComplex ? 'libSQL' : 'DuckDB';
  const ratio4 = Math.max(libsqlComplex, duckdbComplex) / Math.min(libsqlComplex, duckdbComplex);
  console.log(`${'複雑な集計'.padEnd(20)} ${(libsqlComplex.toFixed(2) + 'ms').padStart(15)} ${(duckdbComplex.toFixed(2) + 'ms').padStart(15)} ${faster4.padEnd(10)} ${ratio4.toFixed(2)}x`);

  // UPDATE
  const libsqlUpdate = libsqlTimes.update * 1000;
  const duckdbUpdate = duckdbTimes.update * 1000;
  const faster5 = libsqlUpdate < duckdbUpdate ? 'libSQL' : 'DuckDB';
  const ratio5 = Math.max(libsqlUpdate, duckdbUpdate) / Math.min(libsqlUpdate, duckdbUpdate);
  console.log(`${'UPDATE'.padEnd(20)} ${(libsqlUpdate.toFixed(2) + 'ms').padStart(15)} ${(duckdbUpdate.toFixed(2) + 'ms').padStart(15)} ${faster5.padEnd(10)} ${ratio5.toFixed(2)}x`);

  console.log(`\n${'='.repeat(60)}`);
  console.log('結論');
  console.log('='.repeat(60));
  console.log('\n【libSQL】');
  console.log('  ✓ トランザクション処理 (OLTP) に最適');
  console.log('  ✓ 小〜中規模データでの高速な読み書き');
  console.log('  ✓ 行指向ストレージで個別レコード操作が得意');
  console.log('  ✓ 組み込みアプリケーションに最適');

  console.log('\n【DuckDB】');
  console.log('  ✓ 分析クエリ (OLAP) に最適');
  console.log('  ✓ 大量データの集計・グループ化が高速');
  console.log('  ✓ カラムナストレージによる効率的なスキャン');
  console.log('  ✓ ウィンドウ関数などの高度な分析機能');
  console.log('  ✓ データサイエンス・BI用途に最適');
}

async function main() {
  console.log('='.repeat(60));
  console.log('libSQL vs DuckDB パフォーマンスベンチマーク');
  console.log('='.repeat(60));

  // データサイズを選択
  console.log('\nベンチマークのデータサイズを選択してください:');
  console.log('  1. 小 (10,000件) - 高速');
  console.log('  2. 中 (50,000件) - 推奨');
  console.log('  3. 大 (100,000件) - 時間がかかります');

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  rl.question('\n選択 (1-3, デフォルト=2): ', async (answer) => {
    rl.close();

    const choice = answer.trim() || '2';
    const dataSizes: { [key: string]: number } = {
      '1': 10000,
      '2': 50000,
      '3': 100000,
    };
    const dataSize = dataSizes[choice] || 50000;

    console.log(`\n${dataSize.toLocaleString()}件のデータでベンチマークを開始します...`);

    try {
      // ベンチマーク実行
      const libsqlTimes = await benchmarkLibSQL(dataSize);
      const duckdbTimes = await benchmarkDuckDB(dataSize);

      // 結果比較
      printComparison(libsqlTimes, duckdbTimes, dataSize);

      console.log('\n\n完了!');
    } catch (error) {
      console.error('\nエラーが発生しました:', error);
      process.exit(1);
    }
  });
}

main();
